internal/elasticsearch/transform/transform.go

```diff
@@ -0,0 +1,548 @@
+			Type:        schema.TypeBool,
+			Description: "When true, deferrable validations are not run upon creation, but rather when the transform is started. This behavior may be desired if the source index does not exist until after the transform is created.",
+			Optional:    true,
+			Default:     false,
+		},
+		"timeout": {
+			Type:         schema.TypeString,
+			Description:  "Period to wait for a response from Elastisearch when performing any management operation. If no response is received before the timeout expires, the operation fails and returns an error. Defaults to `30s`.",
+			Optional:     true,
+			Default:      "30s",
+			ValidateFunc: utils.StringIsDuration,
+		},
+		"enabled": {
+			Type:        schema.TypeBool,
+			Description: "Controls wether the transform is started or stopped. Default is `false` (stopped).",
+			Optional:    true,
+			Default:     false,
+		},
+	}
+
+	utils.AddConnectionSchema(transformSchema)
+
+	return &schema.Resource{
+		Schema:      transformSchema,
+		Description: "Manages Elasticsearch transforms. See: https://www.elastic.co/guide/en/elasticsearch/reference/current/transforms.html",
+
+		CreateContext: resourceTransformCreate,
+		ReadContext:   resourceTransformRead,
+		UpdateContext: resourceTransformUpdate,
+		DeleteContext: resourceTransformDelete,
+	}
+}
+
+func resourceTransformCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	transformName := d.Get("name").(string)
+	id, diags := client.ID(ctx, transformName)
+	if diags.HasError() {
+		return diags
+	}
+
+	transform, err := getTransformFromResourceData(ctx, d, transformName)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	params := models.PutTransformParams{
+		DeferValidation: d.Get("defer_validation").(bool),
+	}
+
+	timeout, err := time.ParseDuration(d.Get("timeout").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	params.Timeout = timeout
+
+	params.Enabled = d.Get("enabled").(bool)
+
+	if diags := elasticsearch.PutTransform(ctx, client, transform, &params); diags.HasError() {
+		return diags
+	}
+
+	d.SetId(id.String())
+	return resourceTransformRead(ctx, d, meta)
+}
+
+func resourceTransformRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+	compId, diags := clients.CompositeIdFromStr(d.Id())
+	if diags.HasError() {
+		return diags
+	}
+
+	transformName := compId.ResourceId
+	if err := d.Set("name", transformName); err != nil {
+		return diag.FromErr(err)
+	}
+
+	transform, diags := elasticsearch.GetTransform(ctx, client, &transformName)
+	if transform == nil && diags == nil {
+		tflog.Warn(ctx, fmt.Sprintf(`Transform "%s" not found, removing from state`, compId.ResourceId))
+		d.SetId("")
+		return diags
+	}
+	if diags.HasError() {
+		return diags
+	}
+
```

Comment:
We're not actually updating any state fields in the ResourceData here?

Reply:
At the moment, I'd expect the following to not work as expected:

1. Create a transform via TF
2. Update the transform outside of TF via the ES API directly
3. Re-run the TF plan/apply

Expected:
TF corrects the drift

IIUC, currently TF won't see the configuration drift, since we're not updating the current state in the function. In that case, the plan will come back clean (incorrectly) and no changes will be applied.

Reply:
Indeed, I initially misunderstood the role of the "read" callback; I also got tricked by the fact that the initial tests with a locally-built provider (using the provider_installation / filesystem_mirror config) seemed to work fine for create/update scenarios.
I've added "drift detection" in the todo list.